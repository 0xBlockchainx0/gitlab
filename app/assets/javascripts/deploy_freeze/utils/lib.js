import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';

export const addTimezoneIdentifier = timezoneList => freezePeriod =>
  convertObjectPropsToCamelCase({
    ...freezePeriod,
    cron_timezone: timezoneList.find(tz => tz.identifier === freezePeriod.cron_timezone)?.name,
  });

export default {
  addTimezoneIdentifier,
};
