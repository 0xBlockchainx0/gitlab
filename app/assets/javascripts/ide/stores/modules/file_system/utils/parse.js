/* eslint-disable import/prefer-default-export */
import { insertIntoTree } from './tree';
import { splitPath } from './path';

const TYPE_BLOB = 'blob';
const TYPE_TREE = 'tree';

const ROOT_PATH = '';

const createBaseFileObject = (type, { path = ROOT_PATH, name = ROOT_PATH }) => {
  return {
    type,
    timestamp: 0,
    path,
    name,
  };
};

const createBlob = (...args) => {
  return {
    ...createBaseFileObject(TYPE_BLOB, ...args),
    isLoaded: false,
    content: '',
    active: false,
  };
};

const createTree = (...args) => {
  return {
    ...createBaseFileObject(TYPE_TREE, ...args),
    children: [],
    opened: false,
  };
};

const insertEntry = (files, path, type, createFn) => {
  if (!path || files[path]) {
    return;
  }

  const [parentPath, name] = splitPath(path);

  Object.assign(files, {
    [path]: createFn({ path, name }),
  });

  insertEntry(files, parentPath, TYPE_TREE, createTree);

  insertIntoTree(files[parentPath], type, name);
};

export const parseToFileObjects = paths => {
  const files = {};

  const rootTree = createTree({});
  files[ROOT_PATH] = rootTree;

  paths.forEach(path => {
    insertEntry(files, path, TYPE_BLOB, createBlob);
  });

  return files;
};
