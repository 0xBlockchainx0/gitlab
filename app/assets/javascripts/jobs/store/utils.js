/**
 * Adds the line number property
 * @param Object line
 * @param Number lineNumber
 */
export const parseLine = (line = {}, lineNumber) => ({
  ...line,
  lineNumber,
});

/**
 * When a line has `section_header` set to true, we create a new
 * structure to allow to nest the lines that belong to the
 * collpasible section
 *
 * @param Object line
 * @param Number lineNumber
 */
export const parseHeaderLine = (line = {}, lineNumber) => ({
  isClosed: true,
  isHeader: true,
  line: parseLine(line, lineNumber),
  lines: [],
});

/**
 * `section_duration` is sent in the end of the
 * collpasibl;e section
 *
 * Finds the section it belongs to and adds it to the correct
 * object
 *
 * @param Array data
 * @param Object durationLine
 */
export function addDurationToHeader(data, durationLine) {
  data.forEach(el => {
    if (el.line && el.line.section === durationLine.section) {
      el.line.section_duration = durationLine.section_duration;
    }
  });
}

/**
 * Parses the job log content into a structure usable by the template
 *
 * For collaspible lines (section_header = true):
 *    - creates a new array to hold the lines that are collpasible,
 *    - adds a isClosed property to handle toggle
 *    - adds a isHeader property to handle template logic
 *    - adds the section_duration
 * For each line:
 *    - adds the index as lineNumber
 *
 * @param {Array} lines
 * @returns {Array}
 */
export const logLinesParser = (lines = [], lineNumberStart, accumulator = []) =>
  lines.reduce((acc, line, index) => {
    const lineNumber = lineNumberStart ? lineNumberStart + index : index;
    const last = acc[acc.length - 1];

    if (line.section_header) {
      acc.push(parseHeaderLine(line, lineNumber));
    } else if (
      last &&
      last.isHeader &&
      !line.section_duration &&
      line.section === last.line.section
    ) {
      last.lines.push(parseLine(line, lineNumber));
    } else if (line.section_duration) {
      addDurationToHeader(acc, line);
    } else {
      acc.push(parseLine(line, lineNumber));
    }

    return acc;
  }, accumulator);

/**
 * Finds the repeated offset, removes the old one
 * Returns a new object with the updated log without
 * the repeated offset and the last line number.
 *
 * @param Array newLog
 * @param Array oldParsed
 * @returns Object
 *
 */
export const findOffsetAndRemove = (newLog, oldParsed) => {
  const cloneOldLog = [...oldParsed];
  const lastIndex = cloneOldLog.length - 1;
  const last = cloneOldLog[lastIndex];

  const firstNew = newLog[0];
  const parsed = {};

  if((last.offset === firstNew.offset) || (last.line && last.line.offset === firstNew.offset)) {
    cloneOldLog.splice(lastIndex);
    parsed.lastLine = last.lineNumber;
  } else if (last.lines && last.lines.length) {
    const lastNestedIndex = last.lines.length - 1;
    const lastNested = last.lines[lastNestedIndex];
    if (lastNested.offset === firstNew.offset) {
      last.lines.splice(lastNestedIndex);
      parsed.lastLine = lastNested.lineNumber;
    }
  }
  return cloneOldLog;
};

export const findLastLineNumber = (oldLog) => {
  const lastIndex = oldLog.length - 1;

  if (oldLog[lastIndex].lines && oldLog[lastIndex].lines.length) {
    return oldLog[lastIndex].lines.length -1
  } else if (oldLog.length === 1) {
    return 1;
  } else {
    return lastIndex;
  }
}

/**
 * The first line of the new log may already exist
 *
 * Removes the old line if the offset is the same
 * Parses the new incremented log
 *
 * @param Arrary newLog
 * @param Array oldParsed
 * @returns Array
 */
export const updateIncrementalTrace = (newLog, oldParsed = []) => {
  const parsedLog = findOffsetAndRemove(newLog, oldParsed);
  const lineNumber = findLastLineNumber(oldParsed);
  console.log('lineNumber', lineNumber);
  return logLinesParser(newLog, lineNumber, parsedLog);
};

export const isNewJobLogActive = () => gon && gon.features && gon.features.jobLogJson;
