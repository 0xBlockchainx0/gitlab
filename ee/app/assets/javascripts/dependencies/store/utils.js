export const isDependenciesList = data => Array.isArray(data) && data.length > 0;

export const hasReportStatus = data => Boolean(data && typeof data.report_status === 'string');
