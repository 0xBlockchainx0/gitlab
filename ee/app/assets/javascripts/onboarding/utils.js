import STORAGE_KEY from './constants';

/**
 * onboarding_state:
 * activeTour: Number
 * dismissed: Boolean
 */

const ONBOARDING_PROPS_DEFAULTS = {
  activeTourKey: 1,
  dismissed: false,
};

export const resetOnboardingLocalStorage = () => {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(ONBOARDING_PROPS_DEFAULTS));
};

export const getOnboardingLocalStorageState = () => JSON.parse(localStorage.getItem(STORAGE_KEY));

export const updateLocalStorage = updatedProps => {
  let currentState = getOnboardingLocalStorageState();

  if (!currentState) {
    currentState = resetOnboardingLocalStorage();
  }

  const onboardingState = {
    ...currentState,
    ...updatedProps,
  };

  localStorage.setItem(STORAGE_KEY, JSON.stringify(onboardingState));
};
