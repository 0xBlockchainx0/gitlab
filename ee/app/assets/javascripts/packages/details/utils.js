import { TrackingActions } from './constants';

const trackInstallationTabChange = {
  methods: {
    trackInstallationTabChange(tabIndex) {
      const action = tabIndex === 0 ? TrackingActions.INSTALLATION : TrackingActions.REGISTRY_SETUP;
      this.track(action);
    },
  },
};

export default trackInstallationTabChange;
