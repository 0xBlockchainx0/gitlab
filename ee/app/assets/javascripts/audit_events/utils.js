import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { AVAILABLE_TOKEN_TYPES } from './constants';

export const isNumeric = str => {
  return !Number.isNaN(parseInt(str, 10), 10);
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValues = filterValue.filter(value => AVAILABLE_TOKEN_TYPES.includes(value.type));
  let searchQuery = {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
  };

  if (entityValues.length) {
    const {
      type,
      value: { data: id },
    } = entityValues[0] || { type: null, value: { data: null } };

    searchQuery = { ...searchQuery, entity_id: id, entity_type: type };
  }

  return searchQuery;
};
