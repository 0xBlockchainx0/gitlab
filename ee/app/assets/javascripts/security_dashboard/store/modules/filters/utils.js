import { isSubset } from '~/lib/utils/set';
import { ALL } from './constants';

const createSelection = selectionObj => {
  return Object.values(selectionObj).reduce((acc, curr) => {
    return new Set([...acc, ...curr]);
  }, new Set());
};

export const isBaseFilterOption = id => id === ALL;

/**
 * Returns whether or not the given state filter has a valid selection,
 * considering its available options.
 * @param {Object} filter The filter from the state to check.
 * @returns boolean
 */
export const hasValidSelection = ({ selection, options }) =>
  isSubset(selection, new Set(options.map(({ id }) => id)));

/**
 * Takes a filter array and a selected payload.
 * It then either adds or removes that option from the appropriate selected filter.
 * With a few extra exceptions around the `ALL` special case.
 * @param {Array} filters the filters to mutate
 * @param {Object} payload
 * @param {String} payload.optionId the ID of the option that was just selected
 * @param {String} payload.filterId the ID of the filter that the selected option belongs to
 * @returns {Array} the mutated filters array
 */
export const setFilter = (filters, { filterIds, optionIds }) => {
  return filters.map(filter => {
    if (Object.keys(filterIds).some(id => filter.ids[id])) {
      const { selectionObj: selection } = filter;
      let newSelection;

      /* eslint-disable-next-line guard-for-in, no-restricted-syntax */
      for (const [key, value] of Object.entries(optionIds)) {
        if (Array.isArray(value) && !value.length) {
          continue;
        }
        if (value === ALL) {
          newSelection = { ALL: [ALL] };
          break;
        } else if (selection[key]?.includes(value)) {
          newSelection = { ...selection };
          if (newSelection[key].length === 1) {
            delete newSelection[key];
            if (!Object.keys(newSelection).length) {
              newSelection = { ALL: [ALL] };
            }
          } else {
            newSelection[key].splice(newSelection[key].indexOf(value), 1);
          }
        } else {
          if (!newSelection) {
            newSelection = selection.ALL ? {} : { ...selection };
          }
          if (newSelection[key]) {
            newSelection[key] = newSelection[key].concat(value);
          } else {
            newSelection[key] = Array.isArray(value) ? value : [value];
          }
        }
      }

      return {
        ...filter,
        selection: createSelection(newSelection),
        selectionObj: newSelection,
      };
    }
    return filter;
  });
};
