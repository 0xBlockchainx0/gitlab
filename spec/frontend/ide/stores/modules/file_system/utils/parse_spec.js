import { parseToFileObjects } from '~/ide/stores/modules/file_system/utils/parse';

const TEST_DATA = [
  '123.txt',
  'LICENSE',
  'NEW_FILE',
  'NEW_FILE2',
  'README.md',
  'config/temp.config.js',
  'config/test.config.js',
  'lorem/C.md',
  'src/test/ANOTHER_NEW_FILE',
  'test',
];

describe('~/ide/stores/modules/file_system/utils/parse', () => {
  it('should parse', () => {
    expect(parseToFileObjects(TEST_DATA)).toMatchSnapshot();
  });
});
