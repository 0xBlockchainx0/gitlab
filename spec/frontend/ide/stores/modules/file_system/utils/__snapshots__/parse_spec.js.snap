// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`~/ide/stores/modules/file_system/utils/parse should parse 1`] = `
Object {
  "": Object {
    "children": Array [
      Object {
        "name": "config",
        "type": "tree",
      },
      Object {
        "name": "lorem",
        "type": "tree",
      },
      Object {
        "name": "src",
        "type": "tree",
      },
      Object {
        "name": "123.txt",
        "type": "blob",
      },
      Object {
        "name": "LICENSE",
        "type": "blob",
      },
      Object {
        "name": "NEW_FILE",
        "type": "blob",
      },
      Object {
        "name": "NEW_FILE2",
        "type": "blob",
      },
      Object {
        "name": "README.md",
        "type": "blob",
      },
      Object {
        "name": "test",
        "type": "blob",
      },
    ],
    "name": "",
    "opened": false,
    "path": "",
    "timestamp": 0,
    "type": "tree",
  },
  "123.txt": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "123.txt",
    "path": "123.txt",
    "timestamp": 0,
    "type": "blob",
  },
  "LICENSE": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "LICENSE",
    "path": "LICENSE",
    "timestamp": 0,
    "type": "blob",
  },
  "NEW_FILE": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "NEW_FILE",
    "path": "NEW_FILE",
    "timestamp": 0,
    "type": "blob",
  },
  "NEW_FILE2": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "NEW_FILE2",
    "path": "NEW_FILE2",
    "timestamp": 0,
    "type": "blob",
  },
  "README.md": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "README.md",
    "path": "README.md",
    "timestamp": 0,
    "type": "blob",
  },
  "config": Object {
    "children": Array [
      Object {
        "name": "temp.config.js",
        "type": "blob",
      },
      Object {
        "name": "test.config.js",
        "type": "blob",
      },
    ],
    "name": "config",
    "opened": false,
    "path": "config",
    "timestamp": 0,
    "type": "tree",
  },
  "config/temp.config.js": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "temp.config.js",
    "path": "config/temp.config.js",
    "timestamp": 0,
    "type": "blob",
  },
  "config/test.config.js": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "test.config.js",
    "path": "config/test.config.js",
    "timestamp": 0,
    "type": "blob",
  },
  "lorem": Object {
    "children": Array [
      Object {
        "name": "C.md",
        "type": "blob",
      },
    ],
    "name": "lorem",
    "opened": false,
    "path": "lorem",
    "timestamp": 0,
    "type": "tree",
  },
  "lorem/C.md": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "C.md",
    "path": "lorem/C.md",
    "timestamp": 0,
    "type": "blob",
  },
  "src": Object {
    "children": Array [
      Object {
        "name": "test",
        "type": "tree",
      },
    ],
    "name": "src",
    "opened": false,
    "path": "src",
    "timestamp": 0,
    "type": "tree",
  },
  "src/test": Object {
    "children": Array [
      Object {
        "name": "ANOTHER_NEW_FILE",
        "type": "blob",
      },
    ],
    "name": "test",
    "opened": false,
    "path": "src/test",
    "timestamp": 0,
    "type": "tree",
  },
  "src/test/ANOTHER_NEW_FILE": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "ANOTHER_NEW_FILE",
    "path": "src/test/ANOTHER_NEW_FILE",
    "timestamp": 0,
    "type": "blob",
  },
  "test": Object {
    "active": false,
    "content": "",
    "isLoaded": false,
    "name": "test",
    "path": "test",
    "timestamp": 0,
    "type": "blob",
  },
}
`;
